server/src/main/java/org/apache/kafka/server/AssignmentsManager.java

```diff
@@ -202,9 +199,12 @@ public void run() throws Exception {
                     return;
                 }
                 if (existing.timestampNs > timestampNs) {
-                    existing.onComplete();
-                    if (log.isDebugEnabled()) log.debug("Dropping assignment {} because it's older than {}", this, existing);
+                    existing.merge(this);
+                    if (log.isDebugEnabled()) log.debug("Dropping assignment {} because it's older than existing {}", this, existing);
                     return;
+                } else {
+                    this.merge(existing);
```

Comment:
Doesn't this result in the callback handlers for existing being run twice if existing was retrieved from inflight?